src/mbgl/renderer/renderer.cpp

```diff
@@ -56,6 +56,24 @@ std::vector<Feature> Renderer::queryRenderedFeatures(const ScreenBox& box, const
 AnnotationIDs Renderer::queryPointAnnotations(const ScreenBox& box) const {
     RenderedQueryOptions options;
     options.layerIDs = {{ AnnotationManager::PointLayerID }};
+    return queryAnnotations(box, options);
+}
+
+AnnotationIDs Renderer::queryShapeAnnotations(const ScreenBox& box, const std::vector<std::string> layerIDs) const {
+    RenderedQueryOptions options;
+    __block std::vector<std::string> shapeLayerIDs;
+    if (layerIDs.size()) {
+        shapeLayerIDs.reserve(layerIDs.size());
+        for (const auto &layerID : layerIDs) {
+            shapeLayerIDs.push_back(AnnotationManager::ShapeLayerID + layerID);
```

Comment:
`mbgl::Renderer::Impl` should have a method `queryShapeAnnotations()` that figures out all the shape layer IDs on its own. `mbgl::Renderer::Impl` has a `layerImpls` member that lists all the style layers; iterate over that vector to get each layer that has `AnnotationManager::ShapeLayerID` in its ID. In the end, `-[MGLMapView shapeAnnotationTagsInRect:]` shouldn’t have to build a `layerIDs` vector at all.